File: web/api/moderation.py
"""
web/api/moderation.py
Warn, timeout, kick, ban, unban, purge — behind the framework in `web/moderation.py`.

Every handler here is four lines of its own logic wrapped around
`moderation.authorise`, which runs the six checks before anything reaches
Discord. That shape is the point: a handler cannot perform an action without
having called the thing that decides whether it may, because `authorise` is what
returns the actor object the handler then needs.

Three rules on top of the framework:

* **A reason is mandatory** on every action, and it is written into Discord's
  audit log naming the dashboard, so an incident review can tell where an action
  came from and not just who took it.
* **Every action is logged twice** — once in Discord's audit log (by the API
  call itself) and once in the server's own log channel, best-effort, so it
  appears in the same feed as the actions taken from chat.
* **Purge is capped and channel-scoped.** Discord's bulk delete only reaches 14
  days and 100 messages; the endpoint refuses more rather than silently doing
  less, and it never crosses channels.

`GET /moderation` is what the page paints itself with: every action, whether the
admin can run it, whether the *bot* can, and why not. An action nobody can
perform is shown greyed with the reason rather than hidden.
"""

from __future__ import annotations

import datetime as dt
import logging
from typing import TYPE_CHECKING

import discord
from aiohttp import web

from utils import db

from .. import http as H
from .. import moderation as mod

if TYPE_CHECKING:
    from ..app import Dashboard

log = logging.getLogger("NanoBot.dashboard.moderation")

PURGE_FILTERS = ("anyone", "humans", "bots", "nanobot")


def routes(dash: "Dashboard") -> list:
    @H.require_manager
    async def capabilities(request: web.Request) -> web.Response:
        """What this admin can actually do here, and why not where they can't."""
        guild = H.guild_of(request)
        actor = await dash.member_of(request, guild)
        me = guild.me
        return H.ok(
            {
                "actions": mod.describe(
                    me.guild_permissions.value if me else 0,
                    actor.guild_permissions.value if actor else 0,
                ),
                "limits": {
                    "purge_max": mod.MAX_PURGE,
                    "timeout_max": mod.MAX_TIMEOUT_SECONDS,
                    "reason_max": mod.REASON_MAX,
                },
                "purge_filters": [
                    {"key": "anyone", "label": "Anyone"},
                    {"key": "humans", "label": "People only"},
                    {"key": "bots", "label": "Bots only"},
                    {"key": "nanobot", "label": "NanoBot's own messages"},
                ],
                "warn_config": await db.get_warn_config(guild.id),
                "safety": (
                    "Every action here is the same one the Discord command "
                    "performs, with the same permission and role-hierarchy "
                    "checks. Reasons are required and land in the audit log."
                ),
            }
        )

    @H.require_manager
    async def recent(request: web.Request) -> web.Response:
        """A member's warning history — the read that makes a warn decision."""
        guild = H.guild_of(request)
        raw = request.query.get("member")
        if not raw or not raw.isdigit():
            raise H.bad_request("Which member?")
        warnings = await db.get_warnings(guild.id, int(raw))
        return H.ok({"warnings": warnings, "count": len(warnings)})

    # ── Actions ──────────────────────────────────────────────────────────────
    @H.require_manager
    async def warn(request: web.Request) -> web.Response:
        guild = H.guild_of(request)
        body = await H.body_json(request)
        target = await _member(guild, body.get("member"))
        reason = mod.check_reason(body.get("reason"))
        actor = await mod.authorise(dash, request, guild, "warn", target)

        now = dt.datetime.now(dt.timezone.utc)
        count = await db.add_warning(
            guild.id, target.id, reason, str(actor.id), str(actor), now.isoformat()
        )
        await mod.log_action(dash, guild, actor, "warn", target=target, detail=reason)
        # The auto-kick/ban thresholds are the warnings cog's, and firing them
        # from here would be a second implementation of an escalation policy.
        # The count comes back so the page can say how close they are.
        cfg = await db.get_warn_config(guild.id)
        return H.ok(
            {
                "warnings": count,
                "thresholds": {"kick": cfg.get("kick_at"), "ban": cfg.get("ban_at")},
                "member": _member_payload(target),
            }
        )

    @H.require_manager
    async def timeout(request: web.Request) -> web.Response:
        guild = H.guild_of(request)
        body = await H.body_json(request)
        target = await _member(guild, body.get("member"))
        reason = mod.check_reason(body.get("reason"))
        seconds = _duration(body.get("seconds"))
        actor = await mod.authorise(dash, request, guild, "timeout", target)

        try:
            await target.timeout(
                dt.timedelta(seconds=seconds), reason=mod.audit_reason(actor, reason)
            )
        except discord.Forbidden:
            raise H.forbidden(
                "Discord refused that — check NanoBot's role sits above theirs."
            ) from None
        except discord.HTTPException as exc:
            raise _discord_error(exc) from None

        await mod.log_action(
            dash,
            guild,
            actor,
            "timeout",
            target=target,
            detail=f"{reason} · {seconds // 60} min",
        )
        return H.ok({"member": _member_payload(target), "seconds": seconds})

    @H.require_manager
    async def kick(request: web.Request) -> web.Response:
        guild = H.guild_of(request)
        body = await H.body_json(request)
        target = await _member(guild, body.get("member"))
        reason = mod.check_reason(body.get("reason"))
        actor = await mod.authorise(dash, request, guild, "kick", target)
        payload = _member_payload(target)

        try:
            await target.kick(reason=mod.audit_reason(actor, reason))
        except discord.Forbidden:
            raise H.forbidden(
                "Discord refused that — check the role hierarchy."
            ) from None
        except discord.HTTPException as exc:
            raise _discord_error(exc) from None

        await mod.log_action(dash, guild, actor, "kick", target=target, detail=reason)
        return H.ok({"member": payload})

    @H.require_manager
    async def ban(request: web.Request) -> web.Response:
        guild = H.guild_of(request)
        body = await H.body_json(request)
        target = await _member(guild, body.get("member"))
        reason = mod.check_reason(body.get("reason"))
        # Days of their messages to remove. Discord's own range; defaulting to 0
        # because deleting somebody's history is a separate decision from
        # removing them, and it should be asked for rather than assumed.
        delete_days = _int(body.get("delete_days", 0), "message days", lo=0, hi=7)
        actor = await mod.authorise(dash, request, guild, "ban", target)
        payload = _member_payload(target)

        try:
            await guild.ban(
                target,
                reason=mod.audit_reason(actor, reason),
                delete_message_days=delete_days,
            )
        except discord.Forbidden:
            raise H.forbidden(
                "Discord refused that — check the role hierarchy."
            ) from None
        except discord.HTTPException as exc:
            raise _discord_error(exc) from None

        await mod.log_action(
            dash,
            guild,
            actor,
            "ban",
            target=target,
            detail=reason
            + (f" · removed {delete_days}d of messages" if delete_days else ""),
        )
        return H.ok({"member": payload, "delete_days": delete_days})

    @H.require_manager
    async def unban(request: web.Request) -> web.Response:
        guild = H.guild_of(request)
        body = await H.body_json(request)
        reason = mod.check_reason(body.get("reason"))
        actor = await mod.authorise(dash, request, guild, "unban", None)
        raw = body.get("member")
        try:
            user_id = int(raw)
        except (TypeError, ValueError):
            raise H.bad_request("Which account?") from None

        try:
            user = await dash.bot.fetch_user(user_id)
            await guild.unban(user, reason=mod.audit_reason(actor, reason))
        except discord.NotFound:
            raise H.not_found("That account isn't banned here.") from None
        except discord.Forbidden:
            raise H.forbidden("NanoBot can't manage bans in this server.") from None
        except discord.HTTPException as exc:
            raise _discord_error(exc) from None

        await mod.log_action(dash, guild, actor, "unban", target=user, detail=reason)
        return H.ok({"member": {"id": str(user_id), "name": str(user)}})

    @H.require_manager
    async def bans(request: web.Request) -> web.Response:
        """The ban list, so unbanning is a pick rather than an id you must find."""
        guild = H.guild_of(request)
        await mod.authorise(dash, request, guild, "unban", None)
        rows = []
        try:
            async for entry in guild.bans(limit=100):
                rows.append(
                    {
                        "id": str(entry.user.id),
                        "name": str(entry.user),
                        "avatar": str(entry.user.display_avatar.url),
                        "reason": entry.reason,
                    }
                )
        except discord.Forbidden:
            raise H.forbidden("NanoBot can't read the ban list here.") from None
        return H.ok({"bans": rows})

    @H.require_manager
    async def purge(request: web.Request) -> web.Response:
        """Delete recent messages in one channel.

        Deliberately the *fast* path only — Discord's bulk delete, 100 messages,
        nothing older than 14 days. `/purge mode:slow` exists in chat behind a
        typed confirmation code because it deletes up to 500 at any age one call
        at a time; that is not something to run from a web form, so it stays in
        Discord where the confirmation lives.
        """
        guild = H.guild_of(request)
        body = await H.body_json(request)
        reason = mod.check_reason(body.get("reason"))
        amount = _int(body.get("amount"), "amount", lo=1, hi=mod.MAX_PURGE)
        only = str(body.get("only") or "anyone")
        if only not in PURGE_FILTERS:
            raise H.bad_request("That isn't one of the filters.")
        actor = await mod.authorise(dash, request, guild, "purge", None)

        channel = _channel(guild, body.get("channel_id"))
        me = guild.me
        perms = channel.permissions_for(me) if me else None
        if not perms or not perms.manage_messages:
            raise H.forbidden(f"NanoBot can't delete messages in #{channel.name}.")

        bot_id = dash.bot.user.id if dash.bot.user else 0

        def keep(message: discord.Message) -> bool:
            if only == "humans":
                return not message.author.bot
            if only == "bots":
                return message.author.bot
            if only == "nanobot":
                return message.author.id == bot_id
            return True

        try:
            deleted = await channel.purge(
                limit=amount, check=keep, reason=mod.audit_reason(actor, reason)
            )
        except discord.Forbidden:
            raise H.forbidden(
                f"NanoBot can't delete messages in #{channel.name}."
            ) from None
        except discord.HTTPException as exc:
            raise _discord_error(exc) from None

        await mod.log_action(
            dash,
            guild,
            actor,
            "purge",
            detail=f"{len(deleted)} message(s) in #{channel.name} · {reason}",
        )
        return H.ok(
            {
                "deleted": len(deleted),
                "channel": {"id": str(channel.id), "name": channel.name},
                "note": (
                    "Discord's bulk delete only reaches messages from the last "
                    "14 days, so older ones were skipped."
                    if len(deleted) < amount
                    else None
                ),
            }
        )

    return [
        web.get("/api/guilds/{guild_id}/moderation", capabilities),
        web.get("/api/guilds/{guild_id}/moderation/warnings", recent),
        web.get("/api/guilds/{guild_id}/moderation/bans", bans),
        web.post("/api/guilds/{guild_id}/moderation/warn", warn),
        web.post("/api/guilds/{guild_id}/moderation/timeout", timeout),
        web.post("/api/guilds/{guild_id}/moderation/kick", kick),
        web.post("/api/guilds/{guild_id}/moderation/ban", ban),
        web.post("/api/guilds/{guild_id}/moderation/unban", unban),
        web.post("/api/guilds/{guild_id}/moderation/purge", purge),
    ]


# ── helpers ───────────────────────────────────────────────────────────────────
async def _member(guild: discord.Guild, raw) -> discord.Member:
    try:
        uid = int(raw)
    except (TypeError, ValueError):
        raise H.bad_request("Pick a member.") from None
    member = guild.get_member(uid)
    if member is None:
        try:
            member = await guild.fetch_member(uid)
        except Exception:
            raise H.not_found("They're not in this server.") from None
    return member


def _member_payload(member) -> dict:
    return {
        "id": str(member.id),
        "name": getattr(member, "display_name", str(member)),
        "handle": getattr(member, "name", str(member)),
        "avatar": str(member.display_avatar.url),
    }


def _channel(guild: discord.Guild, raw) -> discord.TextChannel:
    try:
        cid = int(raw)
    except (TypeError, ValueError):
        raise H.bad_request("Pick a channel.") from None
    channel = next((c for c in guild.text_channels if c.id == cid), None)
    if channel is None:
        raise H.bad_request("That channel isn't in this server any more.")
    return channel


def _duration(raw) -> int:
    """Discord's own bounds: a minute at least, 28 days at most."""
    return _int(raw, "duration", lo=60, hi=mod.MAX_TIMEOUT_SECONDS)


def _int(value, field: str, *, lo: int, hi: int) -> int:
    try:
        number = int(value)
    except (TypeError, ValueError):
        raise H.bad_request(f"“{field}” has to be a whole number.") from None
    if not lo <= number <= hi:
        raise H.bad_request(f"“{field}” has to be between {lo:,} and {hi:,}.")
    return number


def _discord_error(exc: discord.HTTPException) -> H.ApiError:
    """Surface Discord's own complaint rather than a generic failure."""
    log.warning("Discord refused a moderation action: %s", exc)
    return H.ApiError(
        502,
        "discord_error",
        f"Discord refused that: {getattr(exc, 'text', None) or exc}",
    )

File: web/static/assets/js/views/moderation.js
/**
 * views/moderation.js
 * Warn, timeout, kick, ban, unban, purge — deliberately the soberest page here.
 *
 * The rest of the dashboard is built to reduce clicks. This one isn't. Every
 * destructive action opens a sheet that makes you pick a member, type a reason,
 * and read what will happen before the button is live; ban and purge add a
 * second confirmation on top. That friction is the feature.
 *
 * Three things it does that a chat command can't:
 *
 *   **It says what you can't do, and why.** Actions the bot or you lack the
 *   permission for are shown greyed with the reason, rather than hidden or —
 *   worse — offered and then refused.
 *   **It shows the history before the decision.** Picking somebody to warn
 *   loads their warning list first, because "third strike" is a different
 *   decision from "first".
 *   **It says where the record goes.** The reason lands in Discord's audit log
 *   naming the dashboard, and the action is echoed into the server's log
 *   channel, so nothing here is quieter than the same action from chat.
 */

import { h, fill } from "../core/dom.js";
import * as api from "../core/api.js";
import * as fmt from "../core/format.js";
import * as ui from "../core/ui.js";

const TIMEOUTS = [
  { value: 60, label: "1 minute" },
  { value: 300, label: "5 minutes" },
  { value: 3600, label: "1 hour" },
  { value: 21600, label: "6 hours" },
  { value: 86400, label: "1 day" },
  { value: 604800, label: "1 week" },
];

export async function moderation({ guildId }) {
  const base = `/api/guilds/${guildId}/moderation`;
  const state = await api.get(base);
  const host = h("div", { class: "stack" });

  const spec = (key) => state.actions.find((a) => a.key === key);

  function paint() {
    fill(
      host,
      h(
        "div",
        {},
        h("h1", {}, "Moderation"),
        h("div", { class: "small dim" }, "The same actions as the Discord commands, same checks")
      ),

      ui.banner(state.safety, { kind: "info", glyph: "🛡️" }),

      ui.card(
        {},
        ui.cardHead("Actions", { glyph: "⚖️" }),
        h(
          "div",
          { class: "grid grid--2" },
          ...state.actions.map((action) =>
            h(
              "button",
              {
                class: "action-tile",
                type: "button",
                disabled: !action.available,
                onClick: () => open(action.key),
              },
              h(
                "span",
                { class: "action-tile__name" },
                action.label,
                action.destructive ? ui.pill("Destructive", "warn") : null
              ),
              h("span", { class: "tiny dim" }, action.available ? action.blurb : action.why_not)
            )
          )
        )
      ),

      thresholdsCard()
    );
  }

  function thresholdsCard() {
    const cfg = state.warn_config;
    if (!cfg.kick_at && !cfg.ban_at) {
      return ui.card(
        {},
        ui.cardHead("Warning thresholds", { glyph: "⚠️" }),
        h(
          "p",
          { class: "small muted", style: { marginBottom: 0 } },
          "No automatic escalation is set up. Warnings are recorded and nothing " +
            "else happens — set thresholds with /warn config in Discord."
        )
      );
    }
    return ui.card(
      {},
      ui.cardHead("Warning thresholds", { glyph: "⚠️", sub: "Applied automatically" }),
      h(
        "div",
        { class: "stack stack--tight small" },
        cfg.kick_at
          ? h("div", {}, `Kicked at ${fmt.plural(cfg.kick_at, "warning")}`)
          : null,
        cfg.ban_at ? h("div", {}, `Banned at ${fmt.plural(cfg.ban_at, "warning")}`) : null,
        h(
          "div",
          { class: "tiny dim" },
          "These fire from the warnings feature itself, so a warning issued here " +
            "escalates exactly as one issued in chat does."
        )
      )
    );
  }

  /* ── Opening an action ────────────────────────────────────────────────── */
  function open(action) {
    if (action === "purge") return openPurge();
    if (action === "unban") return openUnban();
    return openMemberAction(action);
  }

  /**
   * The member-target actions share one sheet: pick, read their history, give a
   * reason, confirm. Sharing it is what keeps the reason field mandatory on all
   * four rather than on the three somebody remembered.
   */
  async function openMemberAction(action) {
    const meta = spec(action);
    const { members } = await api.get(`/api/guilds/${guildId}/members`);
    let target = null;

    const reason = h("textarea", {
      class: "textarea",
      placeholder: "Why? This goes in the audit log.",
      maxlength: String(state.limits.reason_max),
      style: { minHeight: "70px" },
    });
    const seconds = h(
      "select",
      { class: "select" },
      ...TIMEOUTS.map((t) => h("option", { value: t.value, selected: t.value === 3600 }, t.label))
    );
    const deleteDays = h(
      "select",
      { class: "select" },
      h("option", { value: "0", selected: true }, "Keep their messages"),
      h("option", { value: "1" }, "Delete the last day"),
      h("option", { value: "7" }, "Delete the last week")
    );

    const body = h("div", { class: "stack" });
    const search = h("input", {
      class: "input",
      type: "search",
      placeholder: "Search members",
      "aria-label": "Search members",
      onInput: async () => {
        const found = await api.get(
          `/api/guilds/${guildId}/members?q=${encodeURIComponent(search.value)}`
        );
        fill(list, ...found.members.map(memberRow));
      },
    });
    const list = h("div", { class: "stack stack--tight" });
    const detail = h("div", {});

    function memberRow(member) {
      return h(
        "button",
        {
          class: "tile row",
          type: "button",
          style: { width: "100%", cursor: "pointer" },
          onClick: () => pick(member),
        },
        h("img", { class: "avatar avatar--sm", src: member.avatar, alt: "", loading: "lazy" }),
        h("span", { class: "grow", style: { textAlign: "left" } }, member.name),
        h("span", { class: "dim" }, "›")
      );
    }

    async function pick(member) {
      target = member;
      // The history *before* the decision: a third strike is a different call
      // from a first, and having to go and look it up is how it gets skipped.
      const history = await api
        .get(`${base}/warnings?member=${member.id}`)
        .catch(() => ({ warnings: [], count: 0 }));

      fill(
        detail,
        h(
          "div",
          { class: "stack" },
          h(
            "div",
            { class: "row" },
            h("img", { class: "avatar avatar--md", src: member.avatar, alt: "" }),
            h(
              "div",
              { class: "grow" },
              h("strong", {}, member.name),
              h("div", { class: "tiny dim" }, `@${member.handle}`)
            ),
            h(
              "button",
              { class: "btn btn--ghost btn--sm", type: "button", onClick: () => reset() },
              "Change"
            )
          ),
          history.count
            ? h(
                "details",
                { class: "tile" },
                h("summary", { style: { cursor: "pointer" } }, `${fmt.plural(history.count, "previous warning")}`),
                h(
                  "div",
                  { class: "stack stack--tight", style: { marginTop: "var(--s-2)" } },
                  ...history.warnings.slice(-5).reverse().map((w) =>
                    h(
                      "div",
                      { class: "tiny" },
                      h("strong", {}, w.reason),
                      h("span", { class: "dim" }, ` — ${w.by_name}`)
                    )
                  )
                )
              )
            : h("p", { class: "tiny dim" }, "No warnings on record."),
          action === "timeout" ? ui.field("How long", seconds) : null,
          action === "ban" ? ui.field("Their recent messages", deleteDays) : null,
          ui.field("Reason", reason, { hint: "Required. Written to the audit log." }),
          meta.destructive
            ? ui.banner(meta.blurb, { kind: action === "ban" ? "bad" : "warn", glyph: "⚠️" })
            : null
        )
      );
      fill(body, detail);
      view.close();
      openConfirmSheet();
    }

    function reset() {
      target = null;
      fill(body, search, list);
      view.close();
      openPickSheet();
    }

    let view;

    function openPickSheet() {
      fill(body, search, list);
      view = ui.sheet(`${meta.label} someone`, body);
      fill(list, ...members.map(memberRow));
    }

    function openConfirmSheet() {
      view = ui.sheet(`${meta.label} ${target.name}`, body, {
        actions: [
          h("button", { class: "btn", type: "button", onClick: () => view.close() }, "Cancel"),
          ui.actionButton(meta.label, submit, {
            variant: meta.destructive ? "danger" : "primary",
          }),
        ],
      });
    }

    async function submit() {
      if (!reason.value.trim()) {
        ui.warn("Give a reason first — it goes in the audit log.");
        return;
      }
      // Ban gets a second confirmation. It is the one action here that a member
      // cannot walk back from on their own.
      if (action === "ban") {
        const sure = await ui.confirm(
          `Ban ${target.name}?`,
          "They can't rejoin until somebody unbans them. This is written to the " +
            "audit log with your name on it.",
          { danger: true, confirmLabel: "Ban them" }
        );
        if (!sure) return;
      }
      const payload = { member: target.id, reason: reason.value.trim() };
      if (action === "timeout") payload.seconds = Number(seconds.value);
      if (action === "ban") payload.delete_days = Number(deleteDays.value);

      const result = await api.post(`${base}/${action}`, payload);
      view.close();
      if (action === "warn") {
        const near = nearThreshold(result);
        ui.ok(
          `${target.name} warned — that's ${fmt.plural(result.warnings, "warning")}.` +
            (near ? ` ${near}` : ""),
          { title: "⚠️ Warned", ms: near ? 7000 : 4200 }
        );
      } else {
        ui.ok(`${target.name} — done.`, { title: meta.label });
      }
    }

    openPickSheet();
  }

  function nearThreshold(result) {
    const { kick, ban } = result.thresholds || {};
    if (ban && result.warnings >= ban) return "They're at the ban threshold.";
    if (kick && result.warnings >= kick) return "They're at the kick threshold.";
    if (ban && result.warnings === ban - 1) return "One more and they're banned automatically.";
    if (kick && result.warnings === kick - 1) return "One more and they're kicked automatically.";
    return "";
  }

  /* ── Unban ────────────────────────────────────────────────────────────── */
  async function openUnban() {
    const { bans } = await api.get(`${base}/bans`);
    const reason = h("textarea", { class: "textarea", placeholder: "Why are they coming back?" });
    const list = h("div", { class: "stack stack--tight" });

    const view = ui.sheet(
      "Unban somebody",
      h(
        "div",
        { class: "stack" },
        ui.field("Reason", reason, { hint: "Required." }),
        bans.length
          ? list
          : ui.empty("Nobody is banned", "Nothing to undo here.", { glyph: "🎉" })
      )
    );

    fill(
      list,
      ...bans.map((entry) =>
        h(
          "div",
          { class: "tile row" },
          h("img", { class: "avatar avatar--sm", src: entry.avatar, alt: "", loading: "lazy" }),
          h(
            "div",
            { class: "grow" },
            h("div", {}, entry.name),
            entry.reason ? h("div", { class: "tiny dim" }, entry.reason) : null
          ),
          ui.actionButton(
            "Unban",
            async () => {
              if (!reason.value.trim()) {
                ui.warn("Give a reason first.");
                return;
              }
              await api.post(`${base}/unban`, {
                member: entry.id,
                reason: reason.value.trim(),
              });
              view.close();
              ui.ok(`${entry.name} can rejoin now.`);
            },
            { size: "sm" }
          )
        )
      )
    );
  }

  /* ── Purge ────────────────────────────────────────────────────────────── */
  async function openPurge() {
    const { channels } = await api.get(`/api/guilds/${guildId}/channels`);
    const channel = h(
      "select",
      { class: "select" },
      h("option", { value: "" }, "Pick a channel"),
      ...channels.map((c) =>
        h(
          "option",
          { value: c.id, disabled: !c.can_send },
          c.can_send ? `#${c.name}` : `#${c.name} — NanoBot can't act here`
        )
      )
    );
    const amount = h("input", {
      class: "input tabular",
      type: "number",
      min: "1",
      max: String(state.limits.purge_max),
      value: "20",
    });
    const only = h(
      "select",
      { class: "select" },
      ...state.purge_filters.map((f) => h("option", { value: f.key }, f.label))
    );
    const reason = h("textarea", { class: "textarea", placeholder: "Why?" });

    const view = ui.sheet(
      "Purge messages",
      h(
        "div",
        { class: "stack" },
        ui.banner(
          "Deleted messages cannot be recovered. Discord's bulk delete only " +
            "reaches the last 14 days — anything older is skipped rather than " +
            "deleted one at a time.",
          { kind: "bad", glyph: "⚠️" }
        ),
        ui.field("Channel", channel),
        ui.field("How many", amount, {
          hint: `Up to ${state.limits.purge_max} — Discord's limit for a bulk delete.`,
        }),
        ui.field("Whose messages", only),
        ui.field("Reason", reason, { hint: "Required." }),
        h(
          "p",
          { class: "tiny dim" },
          "Need to go further back, or delete more? /purge mode:slow in Discord " +
            "does up to 500 at any age behind a typed confirmation code — which " +
            "is not something to put on a web form."
        )
      ),
      {
        actions: [
          h("button", { class: "btn", type: "button", onClick: () => view.close() }, "Cancel"),
          ui.actionButton(
            "Delete them",
            async () => {
              if (!channel.value) {
                ui.warn("Pick a channel first.");
                return;
              }
              if (!reason.value.trim()) {
                ui.warn("Give a reason first.");
                return;
              }
              const target = channels.find((c) => c.id === channel.value);
              const sure = await ui.confirm(
                `Delete ${amount.value} messages in #${target.name}?`,
                "This can't be undone, and there is no copy anywhere.",
                { danger: true, confirmLabel: "Delete them" }
              );
              if (!sure) return;
              const result = await api.post(`${base}/purge`, {
                channel_id: channel.value,
                amount: Number(amount.value),
                only: only.value,
                reason: reason.value.trim(),
              });
              view.close();
              ui.ok(
                `Deleted ${fmt.plural(result.deleted, "message")} in #${result.channel.name}.` +
                  (result.note ? ` ${result.note}` : ""),
                { title: "🗑️ Purged", ms: result.note ? 7000 : 4200 }
              );
            },
            { variant: "danger" }
          ),
        ],
      }
    );
  }

  paint();
  return host;
}
